### PR TITLE
Enable Mobile Worker Attendees

### PR DESCRIPTION
**Pull request by SmittieC**
## Product Description

This adds a new button on the Attendee Users page that allows web users to select if they want mobile workers to be attendees or not. Enabling this option will create cases of type `commcare-attendee` (or whatever the domain's specific case type for attendees are) for each mobile worker on this domain. Any new mobile worker added while this setting is enabled will also have an associated case.

This PR also allows one to click on the "Add Attendee" button to add an external user as an attendee.

## Technical Summary

This adds a button on the `Attendee Users` page that allows the webuser to select whether or not they want mobile workers as attendees or not. If they enable mobile workers to be attendees, two things will happen:
1. The `AttendanceTrackingConfig.mobile_worker_attendees` flag is updated
2. A background job will run that creates `commcare-attendee` (or whatever the domain's specific case type for attendees are) cases for each mobile worker on this domain.

While the `AttendanceTrackingConfig.mobile_worker_attendees` flag is true, any newly created mobile worker will have an assocaited `commcare-attendee` case. Also, this PR adds a modal which pops up when clicking the add attendee button in order to add an external person as an attendee.

## Feature Flag

Attendance Tracking

### Safety story

- Local testing
- Unit tests

### QA Plan

[QA Ticket](https://dimagi-dev.atlassian.net/browse/QA-4902)

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

#### When the mobile workers can **not** be attendees
Screenshot from 2023-03-15 15-58-09

#### When the mobile workers can be attendees
image

#### Modal for adding an external person as an attendee
image

### CODE REVIEW

File: corehq/apps/events/tasks.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            if user.user_id not in existing:
+                case_blocks.append(get_case_block_for_user(user, user_id))
+            elif existing[user.user_id].closed:
+                closed_cases.append(existing[user.user_id])
+
+        submit_case_blocks(
+            [cb.as_text() for cb in case_blocks],
+            domain=domain_name,
+            user_id=user_id,
+            device_id=__name__,
```
Comment (kaapstorm) [inline corehq/apps/events/tasks.py:40]: We should change this. Maybe to `'corehq.apps.events.tasks.sync_mobile_worker_attendees'`?

File: corehq/apps/events/tasks.py [hunk 2]
```diff
@@ -47,7 +48,7 @@ def get_existing_cases_by_user_ids(domain: str) -> dict[str, CommCareCase]:
 
 
 # TODO: Test
-def get_case_block_for_user(user: CommCareUser, owner_id: str) -> CaseBlock:
+def get_case_block_for_user(user: CommCareUser, owner_id: str, attendee_case_type) -> CaseBlock:
```
Comment (kaapstorm) [inline corehq/apps/events/tasks.py:51]: Alternatively
```suggestion
def get_case_block_for_user(user: CommCareUser, owner_id: str) -> CaseBlock:
    attendee_case_type = get_attendee_case_type(user.domain)
```

---
We will, of course, drop all these useful type hints that allow us to see the properties of `user` immediately. :smirk:  But in a different universe, if we were to keep them, we would apply [this style guide](https://docs.google.com/document/d/1q90SaGuxV9dKnjnEIDdu8ma7xrZJxJq_HsfkCUhQqqk/edit?pli=1#heading=h.78jkbnn4l08n), and give `attendee_case_type` a type hint too:
```suggestion
def get_case_block_for_user(
    user: CommCareUser,
    owner_id: str,
    attendee_case_type: str,
) -> CaseBlock:
```

**Review by kaapstorm — COMMENTED** [review]
Looking good.

**Comment by zandre-eng** [comment]
What would happen if the user clicks on the "Disable Mobile Worker Attendees" button when there are existing mobile workers assigned as attendees? Would these be removed, or do they stay there and just get filtered out? :thinking:

File: corehq/apps/events/tasks.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        reopen_cases(closed_cases)
+
+
+def get_existing_cases_by_user_ids(domain: str) -> dict[str, CommCareCase]:
+    cases = AttendeeCase.objects.by_domain(domain, include_closed=True)
+    return {c.get_case_property(ATTENDEE_USER_ID_CASE_PROPERTY): c for c in cases}
+
+
+# TODO: Test
+def get_case_block_for_user(user: CommCareUser, owner_id: str) -> CaseBlock:
```
[the remark this answers is not in the document]
Comment (SmittieC) [reply corehq/apps/events/tasks.py]: [This PR](https://github.com/dimagi/commcare-hq/pull/32753/files)

**Comment by SmittieC** [comment]
> What would happen if the user clicks on the "Disable Mobile Worker Attendees" button when there are existing mobile workers assigned as attendees? Would these be removed, or do they stay there and just get filtered out? thinking

Good question. I think the associated cases should be closed (Also when downgrading to a non-privileged plan). This functionality I will introduce in a separate PR. I don't want to bloat this one (or the [incoming one](https://github.com/dimagi/commcare-hq/pull/32753/files))

**Comment by Charl1996** [comment]
@SmittieC I wonder if it wouldn't be worthwhile to ask QA to do a pass. That way we also get fresh eyes on the UI and can see if it's as intuitive as we think it is.

**Review by Charl1996 — APPROVED** [review]
I'm happy with these changes, provided QA is also. Lots of it has also already been reviewed by means of "PR-ing" to this branch.

**Pull request merged**